src/NServiceBus.Core/Routing/UnicastRoutingTable.cs

```diff
@@ -46,24 +57,21 @@ public void AddStatic(Type messageType, EndpointInstanceName destination)
         /// <param name="destinationAddress">Destination endpoint instance address.</param>
         public void AddStatic(Type messageType, string destinationAddress)
         {
-            rules.Add((t, c) => StaticRule(t, messageType, new UnicastRoute(destinationAddress)));
+            staticRules.Add((t, c) => StaticRule(t, messageType, new UnicastRoute(destinationAddress)));
         }
 
         /// <summary>
         /// Adds a rule for generating unicast routes.
         /// </summary>
-        /// <param name="dynamicRule">The rule.</param>
-        public void AddDynamic(Func<Type, ContextBag, IEnumerable<IUnicastRoute>> dynamicRule)
+       // /// <param name="dynamicRule">The rule.</param>
+        public void AddDynamic(Func<Type, ContextBag, Task<List<IUnicastRoute>>> dynamicRule)
```

Comment:
Or we could think about integrate caching already by using a similar approach to this

http://blogs.msdn.com/b/pfxteam/archive/2011/01/15/asynclazy-lt-t-gt.aspx

Therefore only once computing the routes and executing the IO bound operation